src/charm.py

```diff
@@ -98,8 +98,13 @@ def on_pebble_ready(self, event: EventBase) -> None:
             event: The event object of the pebble ready event
         """
         container = event.workload
-        config = self._config_file(event)
-        container.push(CONFIG_PATH, config)
+        if not self._stored.agent_config:
+            config = self._config_file(event)
+            container.push(CONFIG_PATH, config)
+            self._stored.agent_config = config
```

Comment:
I don't understand why we are storing the config. We have a method that generates the config file. Why not just call that each time we need it? Is it too slow?

Reply:
Look at [this comment](https://github.com/canonical/grafana-agent-k8s-operator/pull/35#discussion_r799036990)

Reply:
I think the safest way to make this work is to have a method that builds the *whole* config file and call it in both pebble_ready and relation_joined. The method can simply exit if resources are not yet ready and it will be called when the second hook fires.

The reason I think this is better is we have all the information needed to construct the config file statelessly and added a stored state introduces another moving part which could introduce bugs.